### PR TITLE
Replace using genome search endpoints for track panel with thoas data

### PR DESCRIPTION
**Pull request by imransl**
## Type

- [ ] Bug fix
- [ ] New feature
- [x] Improvement to existing feature

## Related JIRA Issue(s)
[ENSWBSITES-1256](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1256)
[ENSWBSITES-1191](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1191)

## Description
There are two tickets covered by this PR.

- ENSWBSITES-1256 replaces the genome search REST endpoints used to retrieve data needed for the track panel list. Thoas data is now used for this purpose.
- ENSWBSITES-1191 shows the MANE Select/Ensembl canonical transcript as thoas supports this

## Deployment URL
http://track-panel-thoas.review.ensembl.org/

## Views affected
Genome browser -> track panel

### CODE REVIEW

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts [hunk 1]
```diff
@@ -16,16 +16,17 @@
...
 import {
   SWISSPROT_SOURCE,
   SPTREMBL_SOURCE
 } from 'src/content/app/entity-viewer/gene-view/components/proteins-list/protein-list-constants';
 
+import { Slice } from 'src/shared/types/thoas/slice';
+import { PhasedExon, Exon } from 'src/shared/types/thoas/exon';
+import { Product, ProductType } from 'src/shared/types/thoas/product';
+import { ExternalReference } from 'src/shared/types/thoas/externalReference';
+import { TranscriptMetadata } from 'ensemblRoot/src/shared/types/thoas/metadata';
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:28]: Notice that when you rely on VSCode to infer your import, it starts the path with `ensemblRoot`. I suppose it pulls it out of `tsconfig.json`. Don't know what to do about this.

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts [hunk 2]
```diff
@@ -191,3 +192,25 @@ export const getProteinXrefs = <
...
+      label: canonical.label,
+      definition: canonical.definition
+    };
+  } else if (mane) {
+    return {
+      label: mane.label,
+      definition: mane.definition
+    };
+  }
+};
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:216]: A very suspicious place for this function. And a very suspicious name for it, in this file (the name made sense in the component file where it was used, but it is confusing here).

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/transcripts-sorter.ts [hunk 3]
```diff
@@ -127,6 +127,8 @@ type GeneViewSortableTranscript = IsProteinCodingTranscriptParam &
   GetSplicedRNALengthParam & {
     slice: SliceWithOnlyLength;
     spliced_exons: unknown[];
+    stable_id?: string;
+    name?: string;
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/helpers/transcripts-sorter.ts:131]: These fields do not take part in the sorting, right? Then they shouldn't be here. This type describes a minimum necessary set of fields to be able to sort the transcripts.

File: src/ensembl/src/gql-client.ts [hunk 4]
```diff
@@ -14,10 +14,11 @@
  * limitations under the License.
  */
 
-import { ApolloClient, InMemoryCache } from '@apollo/client';
+import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
+import fetch from 'cross-fetch';
 
 export const client = new ApolloClient({
-  uri: '/api/thoas',
+  link: new HttpLink({ uri: 'api/thoas', fetch }),
```
Comment (azangru) [inline src/ensembl/src/gql-client.ts:21]: why was this necessary?

File: src/ensembl/src/shared/state/ens-object/ensObjectActions.ts [hunk 5]
```diff
@@ -17,107 +17,152 @@
...
+          }
+          mane {
+            value
+            label
+            ncbi_transcript {
+              id
+              url
+            }
+          }
+          gencode_basic {
```
Comment (azangru) [inline src/ensembl/src/shared/state/ens-object/ensObjectActions.ts:96]: why?

File: src/ensembl/src/shared/types/thoas/transcript.ts [hunk 6]
```diff
@@ -33,4 +33,5 @@ export type FullTranscript = {
   product_generating_contexts: FullProductGeneratingContext[];
   external_references: ExternalReference[];
   metadata: TranscriptMetadata;
+  name?: string;
```
Comment (azangru) [inline src/ensembl/src/shared/types/thoas/transcript.ts:36]: why? we aren't using transcript names anywhere

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts [hunk 7]
```diff
@@ -191,3 +192,25 @@ export const getProteinXrefs = <
...
+    return {
+      label: canonical.label,
+      definition: canonical.definition
+    };
+  } else if (mane) {
+    return {
+      label: mane.label,
+      definition: mane.definition
+    };
+  }
```
Comment (msmanoj) [inline src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:215]: Could be rewritten as:

```javascript
if (mane) {
    return {
      label: mane.label,
      definition: mane.definition
    };
  } else if (canonical) {
    return {
      label: canonical.label,
      definition: canonical.definition
    };
  } 
```

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 2 above)
Comment (msmanoj) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:216]: I agree to that. We already have the metadata when we call this function so there is no point in calling it as `getMetadata`.

I think we could put this back to where it was and then create another function here called `getTranscriptLabelFromMetadata`

File: src/ensembl/src/shared/state/ens-object/ensObjectActions.ts [hunk 8]
```diff
@@ -17,107 +17,152 @@
...
 
-    try {
-      response.track = await apiService.fetch(objectTracksUrl);
-    } catch {
-      // FIXME: this is a temporary solution
-      response.track = builtTrackList(response);
+    if (payload.type === 'region') {
+      const regionObject = buildRegionObject(payload);
+      dispatch(
+        fetchEnsObjectAsyncActions.success({
```
Comment (msmanoj) [inline src/ensembl/src/shared/state/ens-object/ensObjectActions.ts:126]: I think this way of fetching the data is okay but I am not sure if we still need to use `request`, `success` & `failure` actions.

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 7 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:215]: Ah I copied this from `TranscriptQualityLabel.tsx` and didn't bother to check it properly. My bad.

I'll change this in that file too.

File: src/ensembl/src/gql-client.ts (hunk 4 above)
Comment (imransl) [reply src/ensembl/src/gql-client.ts:21]: This is because Node.js doesn't have fetch and we need to pass cross-fetch to make it available for it during SSR

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 1 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:28]: Yeah this is about annoying. The auto import function is quite handy after all, so don't want to disable it either.

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 2 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:216]: @manoj Why do you suggest we put it back rather than just rename this function?

File: src/ensembl/src/shared/state/ens-object/ensObjectActions.ts (hunk 5 above)
Comment (imransl) [reply src/ensembl/src/shared/state/ens-object/ensObjectActions.ts:96]: That got added by mistake. I'll remove it.

File: src/ensembl/src/shared/types/thoas/transcript.ts (hunk 6 above)
Comment (imransl) [reply src/ensembl/src/shared/types/thoas/transcript.ts:36]: I added there since the type is `FullTranscript`. I get your point though. `name` is used only in `ensObjectHelpers.ts`, so I'll extend `FullTranscript` with a new type specifically for it.

File: src/ensembl/src/shared/state/ens-object/ensObjectActions.ts (hunk 8 above)
Comment (imransl) [reply src/ensembl/src/shared/state/ens-object/ensObjectActions.ts:126]: But isn't that the standard we are using at the moment for async actions?

File: src/ensembl/src/shared/types/thoas/transcript.ts (hunk 6 above)
Comment (imransl) [reply src/ensembl/src/shared/types/thoas/transcript.ts]: ✅

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 7 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts]: ✅

File: src/ensembl/src/shared/state/ens-object/ensObjectActions.ts (hunk 5 above)
Comment (imransl) [reply src/ensembl/src/shared/state/ens-object/ensObjectActions.ts]: ✅

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/transcripts-sorter.ts (hunk 3 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/transcripts-sorter.ts]: ✅

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 2 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:211]: I renamed the function ✅

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 1 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts]: ✅

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 2 above)
Comment (azangru) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:211]: > Why do you suggest we put it back rather than just rename this function?

This is a function for a very special presentation concern — displaying a transcript label. Is it possibly a React component?

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx [hunk 9]
```diff
@@ -91,6 +91,9 @@ const QUERY = gql`
           }
         }
         metadata {
+          biotype {
+            value
+          }
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx:96]: Could you remind me why this was necessary? It's in EntityViewer, and neither ENSWBSITES-1191 nor ENSWBSITES-1256 explains this change.

File: src/ensembl/src/shared/state/ens-object/ensObjectHelpers.ts [hunk 10]
```diff
@@ -109,5 +121,69 @@ export const buildRegionObject = (
   };
 };
 
+export const buildEnsObject = (
```
Comment (azangru) [inline src/ensembl/src/shared/state/ens-object/ensObjectHelpers.ts:124]: Currently, ensObjects can be either of type gene or of type region. This function builds an object of a gene type. It looks like it deserves a better name than this.

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts (hunk 2 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/entity-helpers.ts:211]: @azangru This function is reused in `ensObjectHelpers.ts` as it wants to extract the same thing from metadata: https://github.com/Ensembl/ensembl-client/pull/552/files#diff-620ee37acbb79360a848bc6acc9266eccde25c753f4edb74f70081d1df044eceR161

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx (hunk 9 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx:96]: That's something I forgot to remove. I'll get rid of it.

File: src/ensembl/src/shared/state/ens-object/ensObjectHelpers.ts (hunk 10 above)
Comment (imransl) [reply src/ensembl/src/shared/state/ens-object/ensObjectHelpers.ts]: Renamed this to `buildGeneObject` ✅

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx (hunk 9 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx]: ✅

**Comment by azangru** [comment]
Superseded by https://github.com/Ensembl/ensembl-client/pull/563
